proto/pulumi/provider.proto

```diff
@@ -363,81 +373,135 @@ message CheckFailure {
+// provider configuration and state. Inputs supplied to a [](pulumirpc.ResourceProvider.DiffConfig) call should have
+// been previously checked by a call to [](pulumirpc.ResourceProvider.CheckConfig); inputs supplied to a
+// [](pulumirpc.ResourceProvider.Diff) call should have been previously checked by a call to
+// [](pulumirpc.ResourceProvider.Check).
 message DiffRequest {
-    string id = 1;                     // the ID of the resource to diff.
-    string urn = 2;                    // the Pulumi URN for this resource.
-    google.protobuf.Struct olds = 3;   // the old output values of resource to diff.
+    // The ID of the resource being diffed.
+    string id = 1;
 
-    // the new input values of resource to diff, copied from CheckResponse.inputs.
+    // The URN of the resource being diffed.
+    string urn = 2;
+
+    // The old *output* properties of the resource being diffed.
+    google.protobuf.Struct olds = 3;
+
+    // The new *input* properties of the resource being diffed. These should have been validated by an appropriate call
+    // to [](pulumirpc.ResourceProvider.CheckConfig) or [](pulumirpc.ResourceProvider.Check).
     google.protobuf.Struct news = 4;
 
-    repeated string ignoreChanges = 5; // a set of property paths that should be treated as unchanged.
-    google.protobuf.Struct old_inputs = 6;   // the old input values of the resource to diff.
-    string name = 7; // the Pulumi name for this resource.
-    string type = 8; // the Pulumi type for this resource.
+    // A set of [property paths](property-paths) that should be treated as unchanged.
+    repeated string ignoreChanges = 5;
+
+    // The old *input* properties of the resource being diffed.
+    google.protobuf.Struct old_inputs = 6;
+
+    // The name of the resource being diffed. This must match the name specified by the `urn` field, and is passed so
+    // that providers do not have to implement URN parsing in order to extract the name of the resource.
+    string name = 7;
+
+    // The type of the resource being diffed. This must match the type specified by the `urn` field, and is passed so
+    // that providers do not have to implement URN parsing in order to extract the type of the resource.
+    string type = 8;
 }
 
+// `PropertyDiff` describes the kind of change that occurred to a property during a diff operation. A `PropertyDiff` may
+// indicate that a property was added, deleted, or updated, and may further indicate that the change requires a
+// replacement.
 message PropertyDiff {
-    Kind kind = 1; // The kind of diff asdsociated with this property.
-    bool inputDiff = 2; // The difference is between old and new inputs, not old and new state.
+    // The kind of diff associated with this property.
+    Kind kind = 1;
+
+    // True if and only if this difference represents one between a pair of old and new inputs, as opposed to a pair of
```

Comment:
This needs clarification as it keeps tripping me up as a provider implementer. What is this.. We might need to have an operational explanation since this is non-intuitive. I *think* this affects how the paths in the map[path]PropertyDiff map are interpreted to look up values, are the paths resolved against inputs or states. This would be great to nail right here in the comment.

Reply:
Again, I must confess I don't know -- much of these PRs are just cleaning up comments that are there. I am making notes of things to investigate and subsequently flesh out in #17691